source/Cosmos.HAL2/BlockDevice/Ports/Sata.cs

```diff
@@ -0,0 +1,345 @@
+using System;
+using System.Collections.Generic;
+using Cosmos.Core;
+using Cosmos.Core.Memory.Old;
+using Cosmos.HAL.BlockDevice.Registers;
+using Cosmos.Debug.Kernel;
+
+namespace Cosmos.HAL.BlockDevice.Ports
+{
+    public class SATA : StoragePort
+    {
+        internal static Debugger mSATADebugger = new Debugger("HAL", "SATA");
+
+        public override PortType mPortType => PortType.SATA;
+        public override string mPortName => "SATA";
+        public override uint mPortNumber => mPortReg.mPortNumber;
+
+        public PortRegisters mPortReg;
+        public Core.MemoryGroup.AHCI Mem;
+
+        // Constants
+        public const ulong RegularSectorSize = 512UL;
+
+        // Properties
+        private string mSerialNo;
+        private string mFirmwareRev;
+        private string mModelNo;
+
+        public string SerialNo { get => mSerialNo; }
+        public string FirmwareRev { get => mFirmwareRev; }
+        public string ModelNo { get => mModelNo; }
+
+        public SATA(PortRegisters aSATAPort)
+        {
+            // Check if it is really a SATA Port!
+            if (aSATAPort.mPortType != PortType.SATA || (aSATAPort.CMD & (1U << 24)) != 0)
+            {
+                Console.ForegroundColor = ConsoleColor.DarkRed;
+                Console.Write("[Error]");
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.WriteLine($" 0:{aSATAPort.mPortNumber} is not a SATA port!\n");
```

Comment:
Again I'll throw here instead to write in "red" on the OS console, the OS should intercept exceptions and well probably die a painful death after.
Probably the correct exception type is always IOException in these cases...